File: 231.power-of-two.cpp
/*
 * @lc app=leetcode id=231 lang=cpp
 *
 * [231] Power of Two
 */

// @lc code=start
class Solution
{
public:
    bool isPowerOfTwo(int n)
    {
        bool result = false;
        // 如果 n 是 2 的冪，它的二進位只有一個 1
        // If n is a power of 2, its binary representation has only one '1', ex. 1000(8)
        // 減去 1 之後，會把那個 1 變成 0，而 1 右邊全部變成 1
        // After subtracting 1 from n, that '1' bit becomes '0', and all bits to its right become '1'.
        // n & (n - 1) 會剛好變成 0
        // Therefore, n & (n - 1) will be 0. ex. 1000(8) & 0111(7) = 0

        // if (n > 0)
        // {
        //     result = (n & (n - 1)) == 0;
        // }
        // return result;

        if (n < 0)
            return false;
        while (n % 2 == 0)
        {
            n = n / 2;
        }
        return n == 1;
    }
};
// @lc code=end

File: 231.power-of-two.py
#
# @lc app=leetcode id=231 lang=python
#
# [231] Power of Two
#

# @lc code=start
class Solution(object):
    def isPowerOfTwo(self, n):
        """
        :type n: int
        :rtype: bool
        """
        
        result = False
        # If n is a power of 2, its binary representation has only one '1', ex. 1000(8)
        # After subtracting 1 from n, the '1' bit will become '0', and all bits to its right become '1'
        # Therefore, n & (n-1) will be 0. ex. 1000(8) & 0111(7) = 0 

        if n >0:
            result = n & (n-1) ==0
        
        return result
    
        # if n<=0:
        #     return False
        # # This approach repeatedly divides 'n' by 2 until it's no longer evenly divisible by 2.
        # # If 'n' is a power of two, its only prime factor is 2.
        # while (n % 2==0):
        #     #Keep dividing 'n' by 2 using integer division (//)
        #     # Integer division ensures 'n' remains an integer, preventing floating-point issues.
        #     n = n // 2
        # # After the loop, if the original 'n' was a power of two
        # # it will have been reduced exactly to 1.
        # return n==1
    

        
# @lc code=end

